diff -pruN xen-4.13.0/tools/libxl/libxl_create.c xen-4.13.0.new/tools/libxl/libxl_create.c
--- xen-4.13.0/tools/libxl/libxl_create.c	2020-02-06 11:37:43.964628249 +0300
+++ xen-4.13.0.new/tools/libxl/libxl_create.c	2020-02-06 11:34:57.629786673 +0300
@@ -27,6 +27,36 @@
 
 #include <xen-xsm/flask/flask.h>
 
+/* The format of the string is:
+ * domid,aperture_size,gm_size,fence_size. This means we want the vgt
+ * driver to create a vgt instance for Domain domid with the required
+ * parameters. NOTE: aperture_size and gm_size are in MB.
+ */
+static void domcreate_vgt_instance(libxl__gc *gc, uint32_t domid,
+                              libxl_domain_build_info *b_info)
+{
+    const char *path = "/sys/kernel/vgt/control/create_vgt_instance";
+    FILE *vgt_file;
+    int low_gm = b_info->u.hvm.vgt_low_gm_sz ?: 64;
+    int high_gm = b_info->u.hvm.vgt_high_gm_sz ?: 448;
+    int fence = b_info->u.hvm.vgt_fence_sz ?: 4;
+
+    if ((vgt_file = fopen(path, "w")) == NULL) {
+        LOGD(ERROR, domid, "vGT: fopen failed\n");
+        return;
+    }
+
+    LOGD(INFO, domid, "vGT: fprintf %u,%u,%u,%u,-1,0\n", 
+		    domid, low_gm, high_gm, fence);
+    
+    if (fprintf(vgt_file, "%u,%u,%u,%u,-1,0\n",
+               domid, low_gm, high_gm, fence) < 0)
+        LOGD(ERROR, domid, "vGT: fprintf failed %d\n", errno);
+    
+    fclose(vgt_file);
+    return;
+}
+
 int libxl__domain_create_info_setdefault(libxl__gc *gc,
                                          libxl_domain_create_info *c_info,
                                          const libxl_physinfo *info)
@@ -328,6 +358,7 @@ int libxl__domain_build_info_setdefault(
                     return ERROR_INVAL;
                 }
                 break;
+            case LIBXL_VGA_INTERFACE_TYPE_XENGT:
             case LIBXL_VGA_INTERFACE_TYPE_CIRRUS:
             default:
                 if (b_info->video_memkb == LIBXL_MEMKB_DEFAULT)
@@ -1609,6 +1640,9 @@ static void domcreate_launch_dm(libxl__e
         if ( ret )
             goto error_out;
 
+	if (d_config->b_info.u.hvm.vga.kind == LIBXL_VGA_INTERFACE_TYPE_XENGT)
+            domcreate_vgt_instance(gc, domid, &d_config->b_info);
+
         return;
     }
     case LIBXL_DOMAIN_TYPE_PV:
diff -pruN xen-4.13.0/tools/libxl/libxl_dm.c xen-4.13.0.new/tools/libxl/libxl_dm.c
--- xen-4.13.0/tools/libxl/libxl_dm.c	2020-02-06 11:37:43.964628249 +0300
+++ xen-4.13.0.new/tools/libxl/libxl_dm.c	2020-02-06 11:31:14.063182594 +0300
@@ -1422,6 +1422,9 @@ static int libxl__build_device_model_arg
                 GCSPRINTF("qxl-vga,vram_size_mb=%"PRIu64",ram_size_mb=%"PRIu64,
                 (b_info->video_memkb/2/1024), (b_info->video_memkb/2/1024) ) );
             break;
+        case LIBXL_VGA_INTERFACE_TYPE_XENGT:
+            flexarray_vappend(dm_args, "-vga", "xengt", NULL);
+	     break;
         default:
             LOGD(ERROR, guest_domid, "Invalid emulated video card specified");
             return ERROR_INVAL;
diff -pruN xen-4.13.0/tools/libxl/libxl_domain.c xen-4.13.0.new/tools/libxl/libxl_domain.c
--- xen-4.13.0/tools/libxl/libxl_domain.c	2019-12-17 17:23:09.000000000 +0300
+++ xen-4.13.0.new/tools/libxl/libxl_domain.c	2020-02-06 11:31:14.063182594 +0300
@@ -1073,6 +1073,30 @@ static void domain_destroy_callback(libx
 static void destroy_finish_check(libxl__egc *egc,
                                  libxl__domain_destroy_state *dds);
 
+/* We don't care the return value:
+ * 1) the guest may not be a VGT guest;
+ * 2) normally when a VGT guest shutdown, the ioemu has already tried to
+ * destroy the vgt instance and we shouldn't come here by "xl dest dom_id".
+ * 3) we come here because the ioemu didn't destroy the vgt instance
+ * successfully(e.g., ioemu exits abnormally) or we want to kill the guest by
+ * force while it's running. In this case, we still try our best to destroy
+ * the vgt instance.
+ */
+static void destroy_vgt_instance(int domid)
+{
+    const char *path = "/sys/kernel/vgt/control/create_vgt_instance";
+    FILE *vgt_file;
+
+    if (domid <= 0)
+        return;
+
+    if ((vgt_file = fopen(path, "w")) == NULL)
+        return;
+
+    (void)fprintf(vgt_file, "%d\n", -domid);
+    (void)fclose(vgt_file);
+}
+
 void libxl__domain_destroy(libxl__egc *egc, libxl__domain_destroy_state *dds)
 {
     STATE_AO_GC(dds->ao);
@@ -1308,6 +1332,8 @@ static void devices_destroy_cb(libxl__eg
     }
     libxl__userdata_destroyall(gc, domid);
 
+    destroy_vgt_instance(domid);
+
     libxl__unlock_domain_userdata(lock);
 
     /* Clean up qemu-save and qemu-resume files. They are
diff -pruN xen-4.13.0/tools/libxl/libxl_types.idl xen-4.13.0.new/tools/libxl/libxl_types.idl
--- xen-4.13.0/tools/libxl/libxl_types.idl	2020-02-06 11:37:43.965628332 +0300
+++ xen-4.13.0.new/tools/libxl/libxl_types.idl	2020-02-06 11:31:14.064182678 +0300
@@ -38,6 +38,7 @@ libxl_ms_vm_genid = Builtin("ms_vm_genid
 #
 
 MemKB = UInt(64, init_val = "LIBXL_MEMKB_DEFAULT", json_gen_fn = "libxl__uint64_gen_json")
+VgtInt = UInt(32, init_val = "0UL")
 
 #
 # Constants / Enumerations
@@ -220,6 +221,7 @@ libxl_vga_interface_type = Enumeration("
     (2, "STD"),
     (3, "NONE"),
     (4, "QXL"),
+    (5, "XENGT"),
     ], init_val = "LIBXL_VGA_INTERFACE_TYPE_UNKNOWN")
 
 libxl_vendor_device = Enumeration("vendor_device", [
@@ -595,6 +597,9 @@ libxl_domain_build_info = Struct("domain
                                        ("hdtype",           libxl_hdtype),
                                        ("nographic",        libxl_defbool),
                                        ("vga",              libxl_vga_interface_info),
+                                       ("vgt_low_gm_sz",    VgtInt),
+                                       ("vgt_high_gm_sz",   VgtInt),
+                                       ("vgt_fence_sz",     VgtInt),
                                        ("vnc",              libxl_vnc_info),
                                        # keyboard layout, default is en-us keyboard
                                        ("keymap",           string),
diff -pruN xen-4.13.0/tools/xl/xl_parse.c xen-4.13.0.new/tools/xl/xl_parse.c
--- xen-4.13.0/tools/xl/xl_parse.c	2020-02-06 11:37:43.866620092 +0300
+++ xen-4.13.0.new/tools/xl/xl_parse.c	2020-02-06 11:31:14.064182678 +0300
@@ -2590,6 +2590,8 @@ skip_usbdev:
                 b_info->u.hvm.vga.kind = LIBXL_VGA_INTERFACE_TYPE_NONE;
             } else if (!strcmp(buf, "qxl")) {
                 b_info->u.hvm.vga.kind = LIBXL_VGA_INTERFACE_TYPE_QXL;
+            } else if (!strcmp(buf, "xengt")) {
+                b_info->u.hvm.vga.kind = LIBXL_VGA_INTERFACE_TYPE_XENGT;
             } else {
                 fprintf(stderr, "Unknown vga \"%s\" specified\n", buf);
                 exit(1);
@@ -2605,6 +2607,13 @@ skip_usbdev:
                 exit (1);
         }
 
+        if (!xlu_cfg_get_long(config, "vgt_low_gm_sz", &l, 0))
+            b_info->u.hvm.vgt_low_gm_sz = l;
+        if (!xlu_cfg_get_long(config, "vgt_high_gm_sz", &l, 0))
+            b_info->u.hvm.vgt_high_gm_sz = l;
+        if (!xlu_cfg_get_long(config, "vgt_fence_sz", &l, 0))
+            b_info->u.hvm.vgt_fence_sz = l;
+
         xlu_cfg_replace_string (config, "keymap", &b_info->u.hvm.keymap, 0);
         xlu_cfg_get_defbool (config, "spice", &b_info->u.hvm.spice.enable, 0);
         if (!xlu_cfg_get_long (config, "spiceport", &l, 0))
