diff -ru a/src/xorg-qubes.conf.template b/src/xorg-qubes.conf.template
--- a/src/xorg-qubes.conf.template	2019-12-16 17:56:07.635521645 +0300
+++ b/src/xorg-qubes.conf.template	2019-12-16 17:57:03.726228645 +0300
@@ -1,5 +1,6 @@
 Section "Module"
         Load "fb"
+        Load "glamoregl"
 EndSection
 
 Section "ServerLayout"   
@@ -10,8 +11,9 @@
 
 Section "Device"
         Identifier  "Videocard0"
-        Driver      "dummyqbs"
+        Driver      "dummyegl"
         VideoRam %MEM%
+        Option "Render" "/dev/dri/renderD128"
         Option "GUIDomID" "%GUI_DOMID%"
 EndSection
 
