diff --git a/gui-daemon/xside.c b/gui-daemon/xside.c
index 3e12012..183302d 100644
--- a/gui-daemon/xside.c
+++ b/gui-daemon/xside.c
@@ -273,6 +273,7 @@ static Window mkwindow(Ghandles * g, struct windowdata *vm_window)
     XSizeHints my_size_hints;    /* hints for the window manager */
     Atom atom_label;
     int i;
+    XSetWindowAttributes attr;
 
     my_size_hints.flags = PSize;
     my_size_hints.width = vm_window->width;
@@ -283,12 +284,17 @@ static Window mkwindow(Ghandles * g, struct windowdata *vm_window)
     else
         parent = g->root_win;
     // we will set override_redirect later, if needed
-    child_win = XCreateSimpleWindow(g->display, parent,
+    attr.override_redirect = vm_window->override_redirect;
+    child_win = XCreateWindow(g->display, parent,
                     vm_window->x, vm_window->y,
                     vm_window->width,
                     vm_window->height, 0,
-                    BlackPixel(g->display, g->screen),
-                    WhitePixel(g->display, g->screen));
+		    CopyFromParent,
+                    CopyFromParent,
+                    CopyFromParent,
+                    CWOverrideRedirect, &attr);
+                    //BlackPixel(g->display, g->screen),
+                    //WhitePixel(g->display, g->screen));
     /* pass my size hints to the window manager, along with window
        and icon names */
     (void) XSetStandardProperties(g->display, child_win,

