diff --git a/gui-daemon/xside.c b/gui-daemon/xside.c
index 9b4e0cd..3d6db00 100644
--- a/gui-daemon/xside.c
+++ b/gui-daemon/xside.c
@@ -1179,6 +1179,27 @@ static void process_xevent_configure(Ghandles * g, const XConfigureEvent * ev)
                vm_window->width, vm_window->height);
 }
 
+static void send_fullscreen_event(unsigned type, Ghandles * g, struct windowdata *vm_window)
+{
+    XClientMessageEvent ev;
+
+    memset(&ev, 0, sizeof(ev));
+    ev.type = ClientMessage;
+    ev.display = g->display;
+    ev.window = vm_window->local_winid;
+    ev.message_type = g->wm_state;
+    ev.format = 32;
+    ev.data.l[3] = 1; /* source indication: normal application */
+    ev.data.l[1] = g->wm_state_fullscreen;
+    ev.data.l[2] = 0;
+
+    /* ev.data.l[0]: 1 - add/set property, 0 - remove/unset property */
+    ev.data.l[0] = type;
+    XSendEvent(g->display, g->root_win, False,
+        (SubstructureNotifyMask|SubstructureRedirectMask),
+        (XEvent*) &ev);
+}
+
 /* handle VM message: MSG_CONFIGURE
  * check if we like new dimensions/position and move relevant window */
 static void handle_configure_from_vm(Ghandles * g, struct windowdata *vm_window)
@@ -1247,16 +1268,29 @@ static void handle_configure_from_vm(Ghandles * g, struct windowdata *vm_window)
     }
     if (!conf_changed)
         return;
-    vm_window->width = width;
-    vm_window->height = height;
-    vm_window->x = x;
-    vm_window->y = y;
-    if (vm_window->override_redirect)
-        // do not let menu window hide its color frame by moving outside of the screen
-        // if it is located offscreen, then allow negative x/y
-        force_on_screen(g, vm_window, 0,
+
+    // Try to fullscreen fullsized window (stubdomain gui-agent)
+    if (g->allow_fullscreen &&
+        (int) width == g->root_width &&
+        (int) height == g->root_height)
+            send_fullscreen_event(1, g, vm_window);
+    else if (g->allow_fullscreen &&
+        (int) vm_window->width == g->root_width &&
+        (int) vm_window->height == g->root_height)
+            send_fullscreen_event(0, g, vm_window);
+    else
+    {
+        vm_window->width = width;
+        vm_window->height = height;
+        vm_window->x = x;
+        vm_window->y = y;
+        if (vm_window->override_redirect)
+            // do not let menu window hide its color frame by moving outside of the screen
+            // if it is located offscreen, then allow negative x/y
+            force_on_screen(g, vm_window, 0,
                 "handle_configure_from_vm");
-    moveresize_vm_window(g, vm_window);
+        moveresize_vm_window(g, vm_window);
+    }
 }
 
 /* handle local Xserver event: EnterNotify, LeaveNotify
