diff --git a/gui-agent-qemu/qubes-gui.c b/gui-agent-qemu/qubes-gui.c
index 436a2cd..a89fea6 100644
--- a/gui-agent-qemu/qubes-gui.c
+++ b/gui-agent-qemu/qubes-gui.c
@@ -196,13 +196,14 @@ static void process_pv_resize(QubesGuiState * qs)
     if (qs->log_level > 1)
         fprintf(stderr,
                 "handle resize  w=%d h=%d\n", conf.width, conf.height);
+    send_wmhints(qs);
     write_message(qs->vchan, hdr, conf);
     send_pixmap_grant_refs(qs);
-    send_wmhints(qs);
 }
 
 static void handle_configure(QubesGuiState * qs)
 {
+    struct msg_hdr hdr;
     struct msg_configure r;
     read_data(qs->vchan, (char *) &r, sizeof(r));
     fprintf(stderr,
@@ -211,6 +212,11 @@ static void handle_configure(QubesGuiState * qs)
 
     qs->x = r.x;
     qs->y = r.y;
+    //send configure acknowledge
+    hdr.type = MSG_CONFIGURE;
+    hdr.window = QUBES_MAIN_WINDOW;
+    r.override_redirect = 0;
+    write_message(qs->vchan, hdr, r);
 }
 
 static int is_bitset(unsigned char *keys, int num)
