--- a/Makefile.stubdom	2020-02-21 05:46:53.000000000 +0300
+++ b/Makefile.stubdom	2020-04-17 16:41:29.015692629 +0300
@@ -79,7 +79,7 @@ build/qemu/config.status: build/qemu/.pa
 		--disable-vhost-net \
 		--disable-spice \
 		--disable-guest-agent \
-		--audio-drv-list= \
+		--audio-drv-list=pa \
 		--disable-smartcard \
 		--disable-vnc \
 		--disable-spice \
--- a/rootfs/gen	2020-02-21 05:46:53.000000000 +0300
+++ b/rootfs/gen	2020-04-17 16:41:29.004692583 +0300
@@ -55,6 +55,13 @@ do
     ln -s busybox "$rootfs_dir/bin/$i"
 done
 
+#pulseaudio
+mkdir -p "$rootfs_dir"/lib
+cp dist/lib/pulseaudio/*.so* "$rootfs_dir"/lib/
+cp dist/lib/*.so* "$rootfs_dir"/lib/
+inst dist/usr/bin/pulseaudio /bin/pulseaudio
+inst dist/usr/bin/pactl /bin/pactl
+
 make DESTDIR="$PWD/build/qemu/install" -C build/qemu install
 inst build/qemu/install/bin/qemu-system-i386 /bin/qemu
 inst build/qemu/pc-bios/optionrom/linuxboot.bin /share/qemu/linuxboot.bin
@@ -83,11 +90,11 @@ inst "rootfs/qemu-ifdown" "/etc/qemu-ifd
 inst "rootfs/init" "/init"
 
 mkdir $rootfs_dir/lib.new
-(find $rootfs_dir/{,usr/,usr/local/}lib{,64} -type f -print0 || true) | xargs -0 mv -t $rootfs_dir/lib.new/
+(find $rootfs_dir/{,usr/,usr/local/}lib{,64} -type f -print0 || true) | xargs -0 mv -n -t $rootfs_dir/lib.new/
 rm -rf $rootfs_dir/{,usr/,usr/local/}lib{,64}
 IFS=:
 for d in $LD_LIBRARY_PATH; do
-(find $rootfs_dir/$d -type f -print0 || true) | xargs -0 mv -t $rootfs_dir/lib.new/
+(find $rootfs_dir/$d -type f -print0 || true) | xargs -0 mv -n -t $rootfs_dir/lib.new/
 rm -rf $rootfs_dir/$d
 done
 unset IFS
--- a/rootfs/init	2020-02-21 05:46:53.000000000 +0300
+++ b/rootfs/init	2020-04-20 16:44:02.640964348 +0300
@@ -18,6 +18,12 @@ vm_path=$(xenstore-read "/local/domain/$
 # NUL well
 dm_args=$(xenstore-read $(xenstore-list -p "$vm_path/image/dmargs"|sort))
 
+
+
+export PULSE_CONFIG_PATH=/tmp
+export XDG_CONFIG_HOME='/tmp'; 
+pulseaudio --system --use-pid-file=no --daemonize=no --exit-idle-time=-1 --disable-shm=yes -n -vvv -L "module-vchan-sink" -L "module-native-protocol-unix auth-anonymous=1 auth-cookie-enabled=no socket=/tmp/pa.sock" &
+
 # Extract network parameters and remove them from dm_args
 net_args=$(echo "$dm_args" | sed -n '/^-qubes-net:/p')
 dm_args=$(echo "$dm_args" | sed '/^-qubes-net:/d')
@@ -127,13 +133,15 @@ if xenstore-read device/console/1 >/dev/
         stty raw -echo -echoctl <&4
     fi
 fi
+
 # $dm_args and $kernel are separated with \n to allow for spaces in arguments.
 IFS=$'\n'
 set -f
-qemu -sandbox on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny \
+qemu -device ich9-intel-hda -device hda-output,audiodev=qemuhda -audiodev pa,id=qemuhda,server=unix:/tmp/pa.sock,timer-period=1000 -sandbox on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny \
 -chardev pipe,path=/tmp/qmp/qemu,id=m -mon chardev=m,mode=control \
 -chardev socket,server,nowait,path=/tmp/qemu.qmp,id=m2 -mon chardev=m2,mode=control \
     $dm_args $kernel &
+
 set +f
 unset IFS
 
